core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1250,13 +1251,23 @@ class KafkaApis(val requestChannel: RequestChannel,
             metadataResponseTopic(Errors.INVALID_REPLICATION_FACTOR, topic, true, util.Collections.emptyList())
           else
             topicMetadata
-        } else if (allowAutoTopicCreation && config.autoCreateTopicsEnable) {
+        } else if (!isFetchAllMetadata && allowAutoTopicCreation && config.autoCreateTopicsEnable) {
+          // KAFKA-10606: If this request is to get metadata for all topics, auto topic creation should not be allowed
+          // The special handling is necessary on broker side because allowAutoTopicCreation is hard coded to true
+          // for backward compatibility on client side.
           createTopic(topic, config.numPartitions, config.defaultReplicationFactor)
         } else {
           metadataResponseTopic(Errors.UNKNOWN_TOPIC_OR_PARTITION, topic, false, util.Collections.emptyList())
```

Comment:
Could we not avoid this altogether instead of doing the `filter` later?

Reply:
@ijuma I slightly changed it so the processing is a branch in the middle about `isFetchAllMetadata`.

It's still a `map` and `filter` structure because I don't want to sacrifice the readability, but it's handled in the middle instead of at the returning statement, which I think is more elegant and readable than the original version.

Let me know what you think!

Reply:
Also, I added the assertion of no UNKNOWN_TOPIC_OR_PARTITION into the unit test

Reply:
I think if you use `flatMap`, you don't need the `filter` and it's still readable. Thoughts?

Reply:
@ijuma That's a good point.  I've updated it.